### PR TITLE
vCenter password rotation handling.

### PR DESCRIPTION
**Pull request by deepakkinni**
Handling password rotation in plugin.

Testing Done
Verified Password rotation is detected and processed on Vanilla and WCP setup

Pre-Checkins:
https://container-dp.svc.eng.vmware.com/view/CNS-DP/job/Velero-Pipeline-WCP/191/
https://container-dp.svc.eng.vmware.com/job/Container_Precheck_Velero/670/

Tested backup driver to verify that the newly rotated password is picked.
2020-11-03T00:23:57.358834483Z time="2020-11-03T00:23:57Z" level=info msg="syncSecretByKey: Started Secret processing vmware-system-csi/vsphere-config-secret" controller=BackupDriverController logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/backupdriver/backup_driver_controller_base.go:392"
2020-11-03T00:23:57.439344175Z time="2020-11-03T00:23:57Z" level=info msg="Retrieved k8s secret, vsphere-config-secret" controller=BackupDriverController logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/utils/utils.go:99"
2020-11-03T00:23:57.439372017Z time="2020-11-03T00:23:57Z" level=error msg="Failed to unquote value 720 for key cnsregistervolumes-cleanup-intervalinmin. Just store the original value string" controller=BackupDriverController error="invalid syntax" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/utils/utils.go:136"
2020-11-03T00:23:57.439376439Z time="2020-11-03T00:23:57Z" level=info msg="Successfully retrieved latest vSphere VC credentials." controller=BackupDriverController logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/backupdriver/backup_driver_controller_base.go:405"
2020-11-03T00:23:57.43937968Z time="2020-11-03T00:23:57Z" level=info msg="Successfully retrieved VirtualCenterConfig from parameters." logSource="/go/src/github.com/vmware-tanzu/astrolabe/pkg/common/vsphere/vcenter_utils.go:83"
2020-11-03T00:23:57.439382692Z time="2020-11-03T00:23:57Z" level=info msg="Detected VirtualCenterConfig change." logSource="/go/src/github.com/vmware-tanzu/astrolabe/pkg/ivd/ivd_protected_entity_type_manager.go:331"
2020-11-03T00:23:57.439385563Z time="2020-11-03T00:23:57Z" level=info msg="vslmClient wasn't connected, ignoring" logSource="/go/src/github.com/vmware-tanzu/astrolabe/pkg/common/vsphere/virtual_center.go:293"
2020-11-03T00:23:57.43938844Z time="2020-11-03T00:23:57Z" level=info msg="cnsClient wasn't connected, ignoring" logSource="/go/src/github.com/vmware-tanzu/astrolabe/pkg/common/vsphere/virtual_center.go:332"
2020-11-03T00:23:57.439391191Z time="2020-11-03T00:23:57Z" level=info msg="Client wasn't connected, ignoring" logSource="/go/src/github.com/vmware-tanzu/astrolabe/pkg/common/vsphere/virtual_center.go:235"
2020-11-03T00:23:57.439394252Z time="2020-11-03T00:23:57Z" level=info msg="Successfully unregistered VC sc1-10-182-137-43.eng.vmware.com" logSource="/go/src/github.com/vmware-tanzu/astrolabe/pkg/common/vsphere/virtual_center_manager.go:95"
2020-11-03T00:23:57.439397906Z time="2020-11-03T00:23:57Z" level=info msg="Successfully registered VC \"sc1-10-182-137-43.eng.vmware.com\"" logSource="/go/src/github.com/vmware-tanzu/astrolabe/pkg/common/vsphere/virtual_center_manager.go:79"
2020-11-03T00:23:57.439452183Z time="2020-11-03T00:23:57Z" level=info msg="Done resetting volume.defaultVslmManager" logSource="/go/src/github.com/vmware-tanzu/astrolabe/pkg/common/vsphere/vslm_manager.go:244"
2020-11-03T00:23:57.439458243Z time="2020-11-03T00:23:57Z" level=info msg="Done resetting volume.defaultCnsManager" logSource="/go/src/github.com/vmware-tanzu/astrolabe/pkg/common/vsphere/cns_manager.go:96"

Signed-off-by: Deepak Kinni <dkinni@vmware.com>

### CODE REVIEW

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/307/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/308/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/312/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/319/

File: pkg/cmd/datamgr/cli/server/server.go [hunk 1]
```diff
@@ -156,6 +155,7 @@ type server struct {
 	dataMover             *dataMover.DataMover
 	snapManager           *snapshotmgr.SnapshotManager
 	externalDataMgr       bool
+	vcConfigSecret		  bool
```
Comment (xing-yang) [inline pkg/cmd/datamgr/cli/server/server.go:158]: gofmt?

Can you run gofmt in all the directories that you have made changes?

File: pkg/utils/utils.go [hunk 2]
```diff
@@ -948,3 +948,30 @@ func DeleteSvcSnapshot(svcSnapshotName string, gcSnapshotName string, gcSnapshot
...
+		ns = constants.VCSecretNs
+	}
+	return func(obj interface{}) bool {
+		switch obj.(type) {
+		case  *k8sv1.Secret:
+			incomingSecret := obj.(*k8sv1.Secret)
+			return incomingSecret.Namespace == ns &&
+				incomingSecret.Name == constants.VCSecret
+		default:
+			logger.Debugf("Unrecognized object type %+v found during filtering, ignoring", obj)
```
Comment (xing-yang) [inline pkg/utils/utils.go:972]: Don't print out this object in the logs.  It may contain credentials, etc.  We are not sure about it.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/322/

File: pkg/cmd/datamgr/cli/server/server.go (hunk 1 above)
Comment (deepakkinni) [reply pkg/cmd/datamgr/cli/server/server.go]: done

File: pkg/constants/constants.go [hunk 3]
```diff
@@ -135,8 +135,8 @@ const (
 
 // feature flog constants
 const (
-	VSphereLocalModeFlag        = "local-mode"
-	VSphereLocalModeFeature     = "EnableLocalMode"
+	VSphereLocalModeFlag    = "local-mode"
+	VSphereLocalModeFeature = "EnableLocalMode"
```
[the remark this answers is not in the document]
Comment (deepakkinni) [reply pkg/constants/constants.go:139]: it's the same thing, go fmt fixed the formatting I guess.

**Comment by lintongj** [comment]
Here are some general comments. 

Let me rephrase my understanding about this solution: a secret watcher/controller for VC credential is added to backup-driver and data-manager, respectively. Once any update events of the secret of interest is observed, a corresponding "ReloadConfig" action will be called all the way down to IVD petm to reload the VC credential, and reset clients to VC managed objects, like, vcenterManager, vslmManager and cnsManager, as well as, the gvddk client.

Overall, the solution looks unnecessarily complicated to me. Below are a few of points I would like to bring up. 

1. It should be good enough for us to use the CSI secret about VC credential as the source of truth, at least for this release. Hence, there is no point for us to get in-sync with VC credential directly, which replicates the work of syncing up with VC credential at vSphere CSI driver side.
2. The authentication failure due to VC credential rotation would be pretty uncommon as the default rotation period is 12 hours. It would be much more efficient to follow the pull model(retrieve CSI secret on authentication failure) rather than the push model(keep watch the specific secret and push to reload VC config in astrolabe). 
3. It is not a good idea to maintain or deal with any state in Astrolabe, which, as a data protection library, should be stateless. That being said, all the PETM config information should be passed from top, e.g., snapshot manager or data mover in our case. However, in the pending change, we handle too much on VC connection by registring/unregistering VC while reloading VC config.

Therefore, as far as I can see, we probably can make the solution much simpler by considering the following points. 
1. Change to use the pull model and capture the specific error in both backup-driver and data-manager, we need to clearly figure out the error type thrown on the authentication failure.
2. Once the authentication failure is captured, we just need to initialize a new snapshot manager(or data mover) with up-to-date VC credential and update the member variable of snapshot manager(or data mover) in backup-driver server (or data-manager server). 

In this way, we don't need to make much change in astrolabe side. Meanwhile, we also don't need to introduce the secret watcher for the implementation of push model.

**Review by lintongj — COMMENTED** [review]
Here are some general comments. 

Let me rephrase my understanding about this solution: a secret watcher/controller for VC credential is added to backup-driver and data-manager, respectively. Once any update events of the secret of interest is observed, a corresponding "ReloadConfig" action will be called all the way down to IVD petm to reload the VC credential, and reset clients to VC managed objects, like, vcenterManager, vslmManager and cnsManager, as well as, the gvddk client.

Overall, the solution looks unnecessarily complicated to me. Below are a few of points I would like to bring up. 

1. It should be good enough for us to use the CSI secret about VC credential as the source of truth, at least for this release. Hence, there is no point for us to get in-sync with VC credential directly, which replicates the work of syncing up with VC credential at vSphere CSI driver side.
2. The authentication failure due to VC credential rotation would be pretty uncommon as the default rotation period is 12 hours. It would be much more efficient to follow the pull model(retrieve CSI secret on authentication failure) rather than the push model(keep watch the specific secret and push to reload VC config in astrolabe). 
3. It is not a good idea to maintain or deal with any state in Astrolabe, which, as a data protection library, should be stateless. That being said, all the PETM config information should be passed from top, e.g., snapshot manager or data mover in our case. However, in the pending change, we handle too much on VC connection by registring/unregistering VC while reloading VC config.

Therefore, as far as I can see, we probably can make the solution much simpler by considering the following points. 
1. Change to use the pull model and capture the specific error in both backup-driver and data-manager, we need to clearly figure out the error type thrown on the authentication failure.
2. Once the authentication failure is captured, we just need to initialize a new snapshot manager(or data mover) with up-to-date VC credential and update the member variable of snapshot manager(or data mover) in backup-driver server (or data-manager server). 

In this way, we don't need to make much change in astrolabe side. Meanwhile, we also don't need to introduce the secret watcher for the implementation of push model.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/334/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/343/

File: Makefile [hunk 4]
```diff
@@ -32,7 +32,7 @@ DATAMGR_BIN ?= data-manager-for-plugin
 BACKUPDRIVER_BIN ?= backup-driver
 
 RELEASE_REGISTRY = vsphereveleroplugin
-REGISTRY ?= dpcpinternal
+REGISTRY ?= vsp-docker-local.artifactory.eng.vmware.com
```
Comment (xing-yang) [inline Makefile:35]: This should be a separate commit, not part of this PR.

Comment (lintongj) [reply Makefile:35]: It might not be a good idea to put an internal registry as the default in an open source project.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/344/

File: Makefile (hunk 4 above)
Comment (xing-yang) [reply Makefile:35]: Good point.

**Comment by xing-yang** [comment]
Can you squash your commits into one?

File: Makefile [hunk 5]
```diff
@@ -214,9 +214,10 @@ release:
 ifneq (,$(QUALIFIED_TAG))
 	docker pull $(DATAMGR_IMAGE):$(QUALIFIED_TAG)
 	docker pull $(PLUGIN_IMAGE):$(QUALIFIED_TAG)
-	@docker tag $(DATAMGR_IMAGE):$(QUALIFIED_TAG) $(RELEASE_REGISTRY)/$(BACKUPDRIVER_BIN):$(RELEASE_TAG)
-	@docker tag $(DATAMGR_IMAGE):$(QUALIFIED_TAG) $(RELEASE_REGISTRY)/$(DATAMGR_BIN):$(RELEASE_TAG)
-	@docker tag $(PLUGIN_IMAGE):$(QUALIFIED_TAG) $(RELEASE_REGISTRY)/$(PLUGIN_BIN):$(RELEASE_TAG)
+	docker pull $(BACKUPDRIVER_IMAGE):$(QUALIFIED_TAG)
+	docker tag $(BACKUPDRIVER_IMAGE):$(QUALIFIED_TAG) $(RELEASE_REGISTRY)/$(BACKUPDRIVER_BIN):$(RELEASE_TAG)
+	docker tag $(DATAMGR_IMAGE):$(QUALIFIED_TAG) $(RELEASE_REGISTRY)/$(DATAMGR_BIN):$(RELEASE_TAG)
+	docker tag $(PLUGIN_IMAGE):$(QUALIFIED_TAG) $(RELEASE_REGISTRY)/$(PLUGIN_BIN):$(RELEASE_TAG)
```
Comment (xing-yang) [inline Makefile:220]: This is Lintong's PR that is merged already: https://github.com/vmware-tanzu/velero-plugin-for-vsphere/pull/229/files

Looks like a rebase error.

**Comment by deepakkinni** [comment]
Tracking this in https://github.com/vmware-tanzu/velero-plugin-for-vsphere/pull/233
